### PR TITLE
517 extend object on map click 2

### LINKED ISSUE #517: RPV 7 - Extend Link or Node block on map click

### PR DESCRIPTION
**Pull request by kronken**
Closes #517

### CODE REVIEW

File: src/components/sidebar/routePathView/routePathListTab/RoutePathListLink.tsx [hunk 1]
```diff
@@ -13,6 +13,7 @@ interface IRoutePathListLinkProps {
     routePathStore?: RoutePathStore;
     previousNode?: INode;
     routePathLink: IRoutePathLink;
+    reference: any;
```
Comment (neodyymi) [inline src/components/sidebar/routePathView/routePathListTab/RoutePathListLink.tsx]: `Ref<any>`, or even more specific

File: src/components/sidebar/routePathView/routePathListTab/RoutePathListNode.tsx [hunk 2]
```diff
@@ -13,6 +13,7 @@ import * as s from './routePathListObject.scss';
 
 interface IRoutePathListNodeProps {
     node: INode;
+    reference: any;
```
Comment (neodyymi) [inline src/components/sidebar/routePathView/routePathListTab/RoutePathListNode.tsx]: `Ref<any>`, or even more specific

File: src/components/sidebar/routePathView/routePathListTab/RoutePathListObject.tsx [hunk 3]
```diff
@@ -15,10 +15,7 @@ interface IRoutePathListObjectProps {
     headerTypeName: string;
     id: string;
     objectType: ListObjectType;
-}
-
-interface IRoutePathListObjectState {
-    isExtended: boolean;
+    reference: any;
```
Comment (neodyymi) [inline src/components/sidebar/routePathView/routePathListTab/RoutePathListObject.tsx]: `Ref<any>`, or even more specific

File: src/stores/routePathStore.ts [hunk 4]
```diff
@@ -14,23 +14,30 @@ export enum AddLinkDirection {
...
+
 export class RoutePathStore {
     @observable private _isCreating: boolean;
     @observable private _routePath: IRoutePath|null;
     @observable private _hasUnsavedModifications: boolean;
     @observable private _isGeometryValid: boolean;
     @observable private _neighborRoutePathLinks: IRoutePathLink[];
-    @observable private _highlightedNodes: string[];
-    @observable private _highlightedLinks: string[];
+    @observable private _highlightedObjects: string[];
```
Comment (neodyymi) [inline src/stores/routePathStore.ts]: Why is this an array? When can multiple items be highlighted?

File: src/stores/routePathStore.ts [hunk 5]
```diff
@@ -14,23 +14,30 @@ export enum AddLinkDirection {
     AfterNode,
 }
 
+export enum RoutePathViewTab {
```
Comment (neodyymi) [inline src/stores/routePathStore.ts]: Enum vs string literal type - You decide :)

File: src/components/sidebar/routePathView/routePathListTab/RoutePathLinksTab.tsx [hunk 6]
```diff
@@ -17,41 +18,88 @@ interface IRoutePathLinksTabProps {
...
         });
     }
 
+    private onExtend = () => {
+        const extendedObjects = this.props.routePathStore!.extendedObjects;
+        if (extendedObjects.length === 1) {
+            const id = extendedObjects[0];
+            const item = this.listObjectReferences[id].current;
+            const parentHeight = this.listReference.current!.offsetHeight;
+            this.listReference.current!.scrollTo({
```
Comment (neodyymi) [inline src/components/sidebar/routePathView/routePathListTab/RoutePathLinksTab.tsx]: For me, on chrome, this seems to work as well as
```
item.scrollIntoView({ block:'center', behavior: 'smooth' });
```
The last items on the list seem to not be completely scrolled to on click. Maybe the scrolling happens before the extending?

File: src/components/map/layers/edit/UpsertRoutePathLayer.tsx [hunk 7]
```diff
@@ -37,6 +39,18 @@ class NewRoutePathLayer extends Component<IRoutePathLayerProps, IRoutePathLayerS
...
 
+    private defaultActionOnObjectClick = (id: string) => {
+        // Switch to info tab
+        this.props.routePathStore!.setActiveTab(RoutePathViewTab.List);
+        // Close all extended objects, in order to be able to calculate final height of items
+        this.props.routePathStore!.setExtendedObjects([]);
+        // Set extended object, which will trigger automatic scroll
+        this.props.routePathStore!.setExtendedObjects([id]);
+        // Set highlight
+        this.props.routePathStore!.setHighlightedObjects([id]);
```
[the remark this answers is not in the document]
Comment (kronken) [reply src/components/map/layers/edit/UpsertRoutePathLayer.tsx]: This is neither unexpected nor wanted behaviour. I know that it is happening, but I don't know what a reasonable solution would be.

File: src/components/map/layers/edit/UpsertRoutePathLayer.tsx [hunk 8]
```diff
@@ -69,7 +83,7 @@ class NewRoutePathLayer extends Component<IRoutePathLayerProps, IRoutePathLayerS
                     node, previousRPLink, nextRPLink)
                 : undefined;
 
-        let isHighlighted = this.props.routePathStore!.isNodeHighlighted(node.id);
+        let isHighlighted = this.props.routePathStore!.isObjectHighlighted(node.id);
```
[the remark this answers is not in the document]
Comment (kronken) [reply src/components/map/layers/edit/UpsertRoutePathLayer.tsx]: refactored

File: src/components/map/layers/edit/UpsertRoutePathLayer.tsx [hunk 9]
```diff
@@ -93,12 +112,19 @@ class NewRoutePathLayer extends Component<IRoutePathLayerProps, IRoutePathLayerS
...
             this.props.toolbarStore!.selectedTool &&
             this.props.toolbarStore!.selectedTool!.onRoutePathLinkClick ?
                 this.props.toolbarStore!.selectedTool!.onRoutePathLinkClick!(routePathLink.id)
                 : undefined;
 
+        if (
+            this.props.toolbarStore!.selectedTool === null &&
+            !onRoutePathLinkClick
+        ) {
+            onRoutePathLinkClick = () => this.defaultActionOnObjectClick(routePathLink.id);
```
[the remark this answers is not in the document]
Comment (kronken) [reply src/components/map/layers/edit/UpsertRoutePathLayer.tsx]: done

File: src/components/sidebar/routePathView/routePathListTab/RoutePathListLink.tsx (hunk 1 above)
Comment (kronken) [reply src/components/sidebar/routePathView/routePathListTab/RoutePathListLink.tsx]: done

File: src/components/sidebar/routePathView/routePathListTab/RoutePathListNode.tsx (hunk 2 above)
Comment (kronken) [reply src/components/sidebar/routePathView/routePathListTab/RoutePathListNode.tsx]: done

File: src/components/sidebar/routePathView/routePathListTab/RoutePathListObject.tsx (hunk 3 above)
Comment (kronken) [reply src/components/sidebar/routePathView/routePathListTab/RoutePathListObject.tsx]: done

File: src/stores/routePathStore.ts (hunk 4 above)
Comment (kronken) [reply src/stores/routePathStore.ts]: When we are highlighting nodes from where you can continue building routepath, then there are many nodes highlighted

Comment (kronken) [reply src/stores/routePathStore.ts]: hmm, but then this variable is not used. Changed this now to be singular

File: src/stores/routePathStore.ts (hunk 5 above)
Comment (kronken) [reply src/stores/routePathStore.ts]: Enum :)

File: src/components/map/layers/edit/UpsertRoutePathLayer.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (neodyymi) [reply src/components/map/layers/edit/UpsertRoutePathLayer.tsx]: What do you think if all extended objects were highlighted when nothing else is?

**Pull request merged**